### ISSUE TITLE
[WebSocket] Sec-WebSocket-Key Hashing Incorrectly

### ISSUE THREAD

**Issue by pvzig** [vapor/engine#109]
**Platform: Ubuntu 16.04**
**Toolchain: Swift 3.1-DEVELOPMENT-SNAPSHOT-2017-03-13**
**Engine version: 1.3.11**

On Ubuntu 16.04, [`exchange(requestKey)`](https://github.com/vapor/engine/blob/ade43a6262b6a41a6fa2e017b6cbabb447b1b833/Sources/WebSockets/Connection/WebSockets%2BHash.swift#L31) consistently removes the third character from the end of the key, causing incorrect `invalidSecAcceptHeader` errors.

Correct: `7VcW5v8fw9RuwsCtndDDs+VmQus=`
After hashing: `7VcW5v8fw9RuwsCtndDDs+Vmus=`

Correct: `tlfzFb2mOM86dj/ZWpxF0VCvy6s=`
After hashing: `tlfzFb2mOM86dj/ZWpxF0VCv6s=`

**Comment by tanner0101** [comment]
@pvzig can you provide the message for the hash so i can attempt to reproduce?

**Comment by pvzig** [comment]
Just to clarify, this is for `WebSocket+Client`:

**Request:**
```
▿ Request
- GET /websocket/YmYAPid-O8wWn_py2CoT1LYviI8FQztVna4UI9KVrbMzdw2bz4RclzR45KVkmThkiCZNh0ghLNAtms6QNf1qvLKbDa_Hofh5htMUKkVBp7UGhCGAERpofw4AOLmzfa2DSGkpUm92OzA25vLRmBYiV3-BnVeS0NGkOS9z8hUUxXI= HTTP/1.1
- Headers:
	Connection: Upgrade
	Host: mpmulti-rbbl.slack-msgs.com
	Sec-WebSocket-Key: o5eiVVO+8fz5eWtSFBPp4g==
	Upgrade: websocket
	Sec-WebSocket-Version: 13
	User-Agent: App (Swift) VaporEngine/0.9.0
- Body:
	 #0
  ▿ super: HTTP.Message
    - startLine: "GET /websocket/YmYAPid-O8wWn_py2CoT1LYviI8FQztVna4UI9KVrbMzdw2bz4RclzR45KVkmThkiCZNh0ghLNAtms6QNf1qvLKbDa_Hofh5htMUKkVBp7UGhCGAERpofw4AOLmzfa2DSGkpUm92OzA25vLRmBYiV3-BnVeS0NGkOS9z8hUUxXI= HTTP/1.1"
    ▿ headers: 6 key/value pairs
      ▿ (2 elements)
        ▿ key: Connection
          - key: "Connection"
        - value: "Upgrade"
      ▿ (2 elements)
        ▿ key: Host
          - key: "Host"
        - value: "mpmulti-rbbl.slack-msgs.com"
      ▿ (2 elements)
        ▿ key: Sec-WebSocket-Key
          - key: "Sec-WebSocket-Key"
        - value: "o5eiVVO+8fz5eWtSFBPp4g=="
      ▿ (2 elements)
        ▿ key: Upgrade
          - key: "Upgrade"
        - value: "websocket"
      ▿ (2 elements)
        ▿ key: Sec-WebSocket-Version
          - key: "Sec-WebSocket-Version"
        - value: "13"
      ▿ (2 elements)
        ▿ key: User-Agent
          - key: "User-Agent"
        - value: "App (Swift) VaporEngine/0.9.0"
    ▿ body: HTTP.Body.data
      - data: 0 elements
    - storage: 0 key/value pairs
    - peerAddress: nil
  - method: GET
  ▿ uri: wss://mpmulti-rbbl.slack-msgs.com:443/websocket/YmYAPid-O8wWn_py2CoT1LYviI8FQztVna4UI9KVrbMzdw2bz4RclzR45KVkmThkiCZNh0ghLNAtms6QNf1qvLKbDa_Hofh5htMUKkVBp7UGhCGAERpofw4AOLmzfa2DSGkpUm92OzA25vLRmBYiV3-BnVeS0NGkOS9z8hUUxXI=
    - scheme: "wss"
    - userInfo: nil
    - host: "mpmulti-rbbl.slack-msgs.com"
    ▿ port: Optional(443)
      - some: 443
    - path: "/websocket/YmYAPid-O8wWn_py2CoT1LYviI8FQztVna4UI9KVrbMzdw2bz4RclzR45KVkmThkiCZNh0ghLNAtms6QNf1qvLKbDa_Hofh5htMUKkVBp7UGhCGAERpofw4AOLmzfa2DSGkpUm92OzA25vLRmBYiV3-BnVeS0NGkOS9z8hUUxXI="
    - query: nil
    - fragment: nil
    - rawQuery: nil
  ▿ version: 1.1.0
    - major: 1
    - minor: 1
    - patch: 0
```
**Response:**
```
▿ Response
- HTTP/1.1 101 Switching Protocols
- Headers:
	Connection: Upgrade
	Sec-WebSocket-Accept: tlfzFb2mOM86dj/ZWpxF0VCvy6s=
	X-Via: haproxy-ext67
	Upgrade: websocket
- Body:
	 #0
  ▿ super: HTTP.Message
    - startLine: "HTTP/1.1 101 Switching Protocols"
    ▿ headers: 4 key/value pairs
      ▿ (2 elements)
        ▿ key: Connection
          - key: "Connection"
        - value: "Upgrade"
      ▿ (2 elements)
        ▿ key: Sec-WebSocket-Accept
          - key: "Sec-WebSocket-Accept"
        - value: "tlfzFb2mOM86dj/ZWpxF0VCvy6s="
      ▿ (2 elements)
        ▿ key: X-Via
          - key: "X-Via"
        - value: "haproxy-ext67"
      ▿ (2 elements)
        ▿ key: Upgrade
          - key: "Upgrade"
        - value: "websocket"
    ▿ body: HTTP.Body.data
      - data: 0 elements
    - storage: 0 key/value pairs
    ▿ peerAddress: Optional(HTTP.PeerAddress(forwarded: nil, xForwardedFor: nil, stream: "54.242.116.253:443"))
      ▿ some: HTTP.PeerAddress
        - forwarded: nil
        - xForwardedFor: nil
        - stream: "54.242.116.253:443"
  ▿ version: 1.1.0
    - major: 1
    - minor: 1
    - patch: 0
  - status: HTTP.Status.switchingProtocols
  - onComplete: nil
```

Line 57: accept: `tlfzFb2mOM86dj/ZWpxF0VCvy6s=`
Line 58 expected: `tlfzFb2mOM86dj/ZWpxF0VCv6s=`

**Comment by pvzig** [comment]
Just following up, I believe this isn't related specifically to Vapor/Engine but rather a corelibs-foundation bug in `NSData.base64EncodedString` introduced in Swift 3.1 

Relevant links:
https://github.com/IBM-Swift/Kitura-WebSocket/issues/5
https://bugs.swift.org/browse/SR-4435

**Comment by BrettRToomey** [comment]
@LoganWright Could you look into this, please?

**Comment by pvzig** [comment]
This should be resolved in Swift 3.1.1 although I haven't had time to verify.

**Comment by tanner0101** [comment]
We've moved to our own base64 encoder in Vapor 2.
